### ISSUE TITLE
allow Optional, and Nullable with defaults and initializers

### ISSUE THREAD

**Issue by xenoterracide** [immutables/immutables#1282]
related to https://github.com/immutables/immutables/issues/1276

```
warning: (immutables:untype) @Value.Default on a optional attribute make it lose its special treatment
  @Value.Default
```

I have this

```java
 @Override
  @Value.Default
  default Optional<Path> getConfigDir() {
    return Try.ofSupplier( this::getClassLoader )
      .mapTry( cl -> cl.getResource( "config" ) )
      .filter( Objects::nonNull, t -> new IOException( "unable to find resource 'config'" ) )
      .mapTry( URL::toURI )
      .map( Path::of )
      .map( Optional::of )
      .get();
  }
```

in my test I say

```java
  @Test
  void homeNotPresent() {
    var config
      = ImmutableTestCliConfiguration.builder()
      .classLoader( this.getClass().getClassLoader() )
      .configDir( Optional.empty() )
      .moduleType( "missing" )
      .build();
```

but in some tests I want to exclude `configDir` entirely, and let it resolve to the default.

**Comment by xenoterracide** [comment]
@elucash are patches welcome on this? this is kind of a hight priority for me for testing.

**Comment by elucash** [comment]
`@SuppressWarnings("immutables:untype")` can be placed on package. It's a warning, not an error. This can be better handled by having just `Optional<Path> getConfigDir` and having separate method (instance or static) something akin to `getEffectiveConfigDir(config)` doing  `getConfigDir().orElseGet(fromClasspath).orElse(defaultOne)`, the thing that you do `getClasspath` and `IOExceptions` in default initializer is quite awkward, usually you should not do such a side-effect-rich things there, as somewhat better alternative you can make `@Value.Lazy Path getEffectiveConfigDir() { return getConfigDir().orElseGet(fromClasspath).orElse(defaultOne); }`, that is what I usually do for such cases.

**Comment by xenoterracide** [comment]
right, but originally I was just going to allow it to be nullable and not an optional, I couldn't find a way to make that work at all, it's kind of surprising optional works differently, and that optional uses a setter that takes an optional instead of just a value.

what you're suggesting (with `orelseGet`) would mean a lot more code in some of my tests, depending on the project. Obviously this is part of the test builder pattern I mentioned. Although you're mostly right about this usually being a bad idea, it's also just an example. The same problem would exist if it wasn't using IO, but just an `Integer`, the problem being you can't set it to null if it has another default. I guess you're right that the lazy could work, but that seems like a lot of code for just needing to be able to set things to null, it also seems like it wouldn't solve the problem, because you still couldn't set it to null using the builder.

**Comment by elucash** [comment]
> I would also argue that the setter should not take an Optional, but a null, and then the getter of the built object would convert null to optional in a getter.

`Style.optionalAcceptNullable`

**Comment by xenoterracide** [comment]
fair, but that creates an additional issue for me, since the Optional parameter still exists

```
/home/xeno/IdeaProjects/brix/src/test/java/com/xenoterracide/brix/ConfigLoaderTest.java:19: error: [NullOptional] Passing a literal null to an Optional parameter is almost certainly a mistake. Did you mean to provide an empty Optional?
      .configDir( null )
                  ^
    (see https://errorprone.info/bugpattern/NullOptional)
  Did you mean '.configDir( Optional.empty() )'?
```

**Comment by caleb-cushing0-kr** [comment]
> `@SuppressWarnings("immutables:untype")` can be placed on package. It's a warning, not an error.

no it can't?

```java
@SuppressWarnings("immutables:untype")
package com.foo;
```
is a compiletime error...

funny that this has entered purview again.

**Comment by xenoterracide** [comment]
> `@SuppressWarnings("immutables:untype")` can be placed on package. It's a warning, not an error.

no it can't?

```java
@SuppressWarnings("immutables:untype")
package com.foo;
```
is a compiletime error...

funny that this has entered purview again.

**Comment by rupertchen** [comment]
Perhaps the original author meant on the type definition and not the package. I have also successfully used it on a method.
